### DISCUSSION TITLE
Create nested teams for GenericMappingTools organization

### DISCUSSION THREAD

**Discussion by weiji14** [GenericMappingTools/gmt#4732] [Code development discussion]
The [GenericMappingTools organization](https://github.com/orgs/GenericMappingTools/teams) is growing (Woohoo!), and maybe we should think about organizing the teams a bit more. @seisman suggested at https://github.com/GenericMappingTools/pygmt/issues/828#issuecomment-770449616 to use nested teams (see https://docs.github.com/en/github/setting-up-and-managing-organizations-and-teams/about-teams#nested-teams) for the PyGMT folks, and we're thinking whether to apply it to the wider GMT organization (or not).

### Current state

```
-- core - Developers of the core GMT C project (8 members)
-- python - Core devs of PyGMT (5 members)
-- python-contributors - Contributors to PyGMT (6 members)
-- python-maintainers - Maintainers of Python projects (2 members)
-- sphinx_gmt - Core devs of the GMT sphinx extension (2 members)
```

### Proposed state

Names and structure are up for discussion. Members on the top level has least amount of permissions, and those further down the nesting hierarchy have more permissions (e.g. maintain/admin)

```
 members - All members in the GMT project, includes GMT, GMTmex, GMT.jl, PyGMT, etc
 ├── gmt-contributors - Contributors to GMT C project
 |    ├── gmt-core - Developers of the core GMT C project
 |    └── sphinx_gmt - Core devs of the GMT sphinx extension
 ├── python-contributors - Contributors to PyGMT
 |    ├── python-core - Core devs (administrators) of PyGMT
 |    └── python-maintainers - Maintainers of PyGMT
 └── julia-contributors
```

For permission levels, see https://docs.github.com/en/github/setting-up-and-managing-organizations-and-teams/managing-access-to-your-organizations-repositories

**Comment by weiji14** [comment]
Ok, how about we start restructuring the pygmt team first since it's the messiest bunch :stuck_out_tongue:. Following on from https://github.com/GenericMappingTools/pygmt/pull/1056#discussion_r593853769, maybe do something like this:

```
└── python-contributors - Contributors to PyGMT (write permission)
      ├── python-maintainers - Maintainers of PyGMT (maintain permission)
      └── python - Core developers of PyGMT (admin permission)
```

What this means that every python contributor has 'write' permission as a baseline (instead of 'triage' currently), and mentioning @GenericMappingTools/python-contributors will ping everyone working on the PyGMT project. We *could* have @GenericMappingTools/python nested under @GenericMappingTools/python-maintainers, but I think it might be better put them on the same nesting level so we could ping either the maintainers or the core-devs separately. Thoughts?

**Reply by core-man** [reply]
> What this means that every python contributor has 'write' permission as a baseline (instead of 'triage' currently), and mentioning @GenericMappingTools/python-contributors will ping everyone working on the PyGMT project. 

I think such nested teams are great. When all contributors are in `python-contributors`, contributors only need to ping this team for regular review/comment requests. It is good for every python contributor to have 'write' permission if the contributor regularly contributes to PyGMT. But does this request a new contributor to be familiar with git workflow since we can directly create and write a feature branch on PyGMT repo?

> We could have @GenericMappingTools/python nested under @GenericMappingTools/python-maintainers, but I think it might be better put them on the same nesting level so we could ping either the maintainers or the core-devs separately. Thoughts?

Does it mean that not all core developers belong to `python-maintainer`?  Some core developers only help to review or some other things?

**Reply by weiji14** [reply]
> I think such nested teams are great. When all contributors are in `python-contributors`, contributors only need to ping this team for regular review/comment requests.

Yep, the simplified pinging is definitely the main benefit of nested teams, and the reason for starting this discussion in the first place!

> It is good for every python contributor to have 'write' permission if the contributor regularly contributes to PyGMT. But does this request a new contributor to be familiar with git workflow since we can directly create and write a feature branch on PyGMT repo?

A brand new, first-time contributor will always have to do so from her/his own PyGMT fork. It is only when we add them to the python-contributors team (we typically send an invite when they add their name to AUTHORS.md) that they get access to "write" permissions. The git workflow does have a learning curve, but there's a first time for everyone and it's easier to teach people on the spot :D

> > We could have @GenericMappingTools/python nested under @GenericMappingTools/python-maintainers, but I think it might be better put them on the same nesting level so we could ping either the maintainers or the core-devs separately. Thoughts?
> 
> Does it mean that not all core developers belong to `python-maintainer`? Some core developers only help to review or some other things?

Yes, core developers need not be maintainers (though their admin permission does give them all of the maintain permissions). Up to them what they want to do with that privilege though.

I should note that even though I have admin permissions from the PyGMT python team, there are still some things I can't do (e.g. make new teams or install GitHub apps) since I'm not a GMT code owner, but @seisman and @meghanrjones will be able to do so.

**Reply by core-man** [reply]
> The git workflow does have a learning curve, but there's a first time for everyone and it's easier to teach people on the spot :D

Great. Actually, I begin to be familiar with git workflow when I contribute to PyGMT :smile: . So I think `python-contributors` is okay based on your suggestion.

**Reply by maxrjones** [reply]
I think there may be people who would actually prefer *not* having write permissions after their first PR(s) and would want time to get more experience with the workflow while having the knowledge that they can't accidentally commit to the 'master' branch or something like that. I would like to see an opportunity for people with such preferences to still increase their involvement and get notified of changes, etc. (i.e., a triage team). 

A couple other preferences - calling the teams 'pygmt' rather than 'python' and eventually (not now) we may want to set up a documentation team.

**Reply by weiji14** [reply]
> I think there may be people who would actually prefer _not_ having write permissions after their first PR(s) and would want time to get more experience with the workflow

Yes I agree. New contributors won't get added to the python-contributors team immediately. What typically happens is:

1. They add their name to AUTHORS.md
2. We send them an invite to join the python-contributors team (which is at our discretion, we could withhold giving them access until they demonstrate enough practice with three or more commits under their GitHub account).
3. They have to then accept the invitation to join, and some people may choose not to if they wish

> ... while having the knowledge that they can't accidentally commit to the 'master' branch or something like that.

The 'master' branch is protected against accidental commits (and force pushes, even from me). So rest assured that this won't happen.

![image](https://user-images.githubusercontent.com/23487320/111083891-9fe1ad80-8574-11eb-97b5-e4c0896caa1f.png)

> I would like to see an opportunity for people with such preferences to still increase their involvement and get notified of changes, etc. (i.e., a triage team).
> 
> A couple other preferences - calling the teams 'pygmt' rather than 'python' and eventually (not now) we may want to set up a documentation team.

At this stage (with about 10 people under the PyGMT umbrella), I think we can do with less of a structure, so maybe keep the 'triage' and 'documentation' idea for the future (e.g. when we have 50-100 contributors).

And on renaming 'GenericMappingTools/python' to 'GenericMappingTools/pygmt', what are people's thoughts on that?

**Reply by seisman** [reply]
I agree with @weiji14 that "write" permission is better and more convenient than the "triage" permission for the pygmt-contributors team, but I also agree with @meghanrjones about the concern of "write" permission. If I understand the "write" permission correctly, contributors with write permission can also commit changes to branches that are NOT created by them. It's a little dangerous for people who are new to git or git workflow. 

To balance the convenience and security concerns, I think we could change the criterion about when a contributor should be added to the teams.

As per the [PyGMT authorship](https://github.com/GenericMappingTools/pygmt/blob/master/AUTHORSHIP.md), anyone can add themselves to AUTHORS.md after their first PR is merged. This is a good way to give credits to these contributors. As mentioned by @weiji14, now we usually invite the contributors to the python-contributors team, after they add themselves to the AUTHORS.md team. The problems are:

- It's dangerous if python-contributors have written permission but the contributor are not very familiar with git (as mentioned by @meghanrjones)
- Some contributors may submit their first PR and no longer participant in this project. They don't need write permission and it makes little sense to ping them.
---

Perhaps we can give python-contributors "write" permission, but only add people who make continuous contributions to the project.

So the team structure and permission levels are still like the one @weiji14 proposed (https://github.com/GenericMappingTools/gmt/discussions/4732#discussioncomment-479861), but the process is a little different:

1. Invite contributors to add themselves to the AUTHORS.md after their first PR
2. Invite them to the python-contributors team (with write permission) after they make more PRs and show good git skills (e.g., @core-man)
3. Invite/promote them to the python-maintainer team (with maintaining permission) when they actively submit PRs and participant in discussions (e.g., @willschlitzer)
4. Invite/promote them to the python team (with admin permission) when we feel someone should be given admin permission. (As I mentioned somewhere, we should limit the number of people in this team for security reasons).
 
Sometimes I feel that people (including myself) may not fully understand what it means and what is changed by being added to a team. So, when we invite anyone to the team, we should also let them know:

- after joining the team, they will have write/maintainance/admin permission, and they can do things like "working in the pygmt repo instead of a fork"
- they should be more responsible as they have more permissions
- they should be more cautious about their operations
- we will ping them by `@python-contributors` when we want their comments or help and they will be notified

**Reply by maxrjones** [reply]
One of my concerns with giving the umbrella team (pygmt or python-contributors, whatever the name will be) write permissions is that it will create a mess for team maintenance if we ever want to create a sub-team with lesser permissions (possibly needing to completely recreate all teams). I strongly prefer setting permissions for each sub-team independently rather than giving permissions to the umbrella team.

Edit: my preference would be like this (so everyone could be pinged by python or pygmt- whichever the name is- but there aren't permissions set there):
pygmt
    - pygmt-contributors (write)
    - pygmt-maintainers (maintain)
    - pygmt-admin or pygmt-core (admin)

**Reply by seisman** [reply]
> pygmt
> - pygmt-contributors (write)
> - pygmt-maintainers (maintain)
> - pygmt-admin or pygmt-core (admin)

Sounds good to me. I also prefer "pygmt" to "python".

**Reply by weiji14** [reply]
Lots of good points thanks @seisman for the clearly laid out arguments and @meghanrjones for thinking about the structure long-term! Looks like we'll need an action item to edit https://github.com/GenericMappingTools/pygmt/blob/v0.3.1/MAINTENANCE.md#onboarding-access-checklist!

Just do be clear, do we want a structure like this (Option 1):

```
└── pygmt -  The PyGMT umbrella team (triage, but nobody will be in here until we grow very big)
      ├── pygmt-contributors - (write)
      ├── pygmt-maintainers - (maintain)
      └── pygmt-admin/core - (admin)
```

Or this (Option 2):

```
└── pygmt -  The PyGMT umbrella team (triage, but nobody will be in here until we grow very big)
      ├── pygmt-contributors - (write)
      |  └── pygmt-maintainers - (maintain)
      └── pygmt-admin/core - (admin)
```
Or this (Option 3):
```
└── pygmt -  The PyGMT umbrella team (triage, but nobody will be in here until we grow very big)
      └── pygmt-contributors - (write)
        └── pygmt-maintainers - (maintain)
           └── pygmt-admin/core - (admin)
```
Or something else entirely?

**Reply by seisman** [reply]
As discussed above, most contributors will be added to the pygmt-contributors team first, and some of them will be added to pygmt-maintainers later. I don't see any benefits using nested teams (option 3). I think option 1 is the best.

**Reply by core-man** [reply]
I think option 1 is better. but I have a question about pygmt-contributors. Do we still need to put any pygmt-maintainer or -admin/core in python-contributors? Or when we want to give more permission to a pygmt-contributors, **copy** or **move** to another team? I feel move better since already have a pygmt umbrella team with everyone.  If copy is used, it's better to put pygmt-maintainers and -admin/core within pygmt-contributors. Or we will use both copy and move？

**Reply by maxrjones** [reply]
I also prefer option 1. My preference is for pygmt to be read access (I don't see people being assigned to a team at that level) and for us to create a pygmt-triage sub-team if needed later on. I do not expect this decision to matter, but prefer to leave the option open. Regarding @core-man's question, I agree that moving would probably be simpler than copying.

**Reply by seisman** [reply]
If I understand @core-man correctly, "moving" means that any user only belong to one team (pygmt-contributors, pygmt-maitainers or pygmt-admin). So @meghanrjones @weiji14 and I will be in the pygmt-admin team, and @core-man will be in the pygmt-contributors team. When contributors (including us) submit PRs and want to ping someone for review or help, how do they ping us? They have to ping all the 3 teams. I'm afraid it will cause more confusion.

If we use the "copying" mechanism, then we only need to ping pygmt-contributors for review, and ping pygmt-maintainers for merging the PR, and ping pygnt-admin for rare cases that may need admin permission.

So "nested teams" (Option 3) makes more sense, although the nested structure looks complicated.

**Reply by maxrjones** [reply]
Good points, @seisman and @core-man.

**Reply by core-man** [reply]
@seisman good summary. you got me. it is the question that confuses me since I began to contribute to PyGMT.

I think nested team is only complicated at the begining and we need put all members at the right teams. But when we want to update one member's permission, it is easy (just copy).

Isn't it like Python base and sub classes? :smile:

**Comment by core-man** [comment]
> Perhaps we can give python-contributors "write" permission, but only add people who make continuous contributions to the project.
> 
> So the team structure and permission levels are still like the one @weiji14 proposed ([#4732 (comment)](https://github.com/GenericMappingTools/gmt/discussions/4732#discussioncomment-479861)), but the process is a little different:
> 
> 1. Invite contributors to add themselves to the AUTHORS.md after their first PR
> 2. Invite them to the python-contributors team (with write permission) after they make more PRs and show good git skills (e.g., @core-man)
> 3. Invite/promote them to the python-maintainer team (with maintaining permission) when they actively submit PRs and participant in discussions (e.g., @willschlitzer)

What's the difference between pygmt-maintainers (maintain) and pygmt-admin/core (admin)? Or equally, If pygmt-contributors has the access to write, it seems that there is little difference between pygmt-maintainers and pygmt-contributors. Is my understanding right?

> 4. Invite/promote them to the python team (with admin permission) when we feel someone should be given admin permission. (As I mentioned somewhere, we should limit the number of people in this team for security reasons).

Yes, I also think so. **The admin permission is very important and should have a limited number.**  Most contributors do not need admin permission to help with PyGMT development. I'd like to suggest PyGMT should only give admin permission to some contributors who continuously contribute to PyGMT, have some good skills (e.g., git, teamwork), win the trust from pygmt-admin and/or other contributors, and (most important) have the passion and responsibility to maintain PyGMT. Meanwhile, when we have enough active administration (e.g., 3~4?), we should not increase pygmt-admin.

> Sometimes I feel that people (including myself) may not fully understand what it means and what is changed by being added to a team. So, when we invite anyone to the team, we should also let them know:
> 
> * after joining the team, they will have write/maintainance/admin permission, and they can do things like "working in the pygmt repo instead of a fork"
> * they should be more responsible as they have more permissions
> * they should be more cautious about their operations
> * we will ping them by `@python-contributors` when we want their comments or help and they will be notified

Great. More permissions, more responsibility. Should be more cautious. I think we should let all the contributors/maintenance/administration know it.

**Reply by seisman** [reply]
> What's the difference between pygmt-maintainers (maintain) and pygmt-admin/core (admin)? Or equally, If pygmt-contributors has the access to write, it seems that there is little difference between pygmt-maintainers and pygmt-contributors. Is my understanding right?

They're different. See this long table (https://docs.github.com/en/github/setting-up-and-managing-organizations-and-teams/repository-permission-levels-for-an-organization#repository-access-for-each-permission-level) for details.

The most important differences are:

- pygmt-contributors (with write permission) can create branches in the PyGMT repository
- pygmt-maintainers (with maintenance permission) can approve and merge a branch into master (python-contributors can't)
- pygmt-admin can make any changes and maybe can delete the whole repository 😃

**Reply by core-man** [reply]
> * pygmt-maintainers (with maintenance permission) can approve and merge a branch into master (python-contributors can't)

See https://github.com/GenericMappingTools/pygmt/pull/1047#issuecomment-798888150.  @willschlitzer cannot merge that PR due to the required test.

> > Let's get this merged! Zenodo is at the ready too!
> 
> I can't "Squash and merge" because of the required tests; can you disable them/merge this PR?

**Reply by seisman** [reply]
That's because some CI status checks didn't pass (sometimes tests crash on Windows).

**Reply by core-man** [reply]
Then I think Option 3 is better. For this nested team,  it seems that `pygmt-contributors` (for review/comments) and `pygmt-maintainers` (for maintenance/merging PR) will be mentioned for most cases during regular PyGMT development and maintenance, while `pygmt-admin/core` (see a case in https://github.com/GenericMappingTools/pygmt/pull/1047#issuecomment-798888150) and `pygmt` teams will only be used for some special cases. This sounds reasonable to me.

**Reply by weiji14** [reply]
Can we give Option 2 some consideration? I'm just thinking of people like Leo and Liam who are currently admins, but probably don't want to be pinged for day-to-day maintenance tasks. Choosing Option 2 will mean we duplicate some people in both maintenance and admin (e.g. @seisman and me) but I think it's a nice middle ground.

**Reply by core-man** [reply]
It seems that the substructure below is what we want now.
> ```
> └── pygmt -  The PyGMT umbrella team (triage, but nobody will be in here until we grow very big)
>       ├── pygmt-contributors - (write)
>       |  └── pygmt-maintainers - (maintain)
> ```

Meanwhile, we could actually choose to put `pygmt-admin/core` either under `pygmt`, `pygmt-contributors`, or `pygmt-maintainer`. For the reason @weiji14  mention above,  it may be better to put it under `pygmt` since I think `pygmt-contributors` and `pygmt-maintainers` will be mentioned during the long-term PyGMT development.

We can also move some maintainers back to `pygmt-contributors` if they don't want to be mentioned for daily maintenance tasks, and contributors back to `pygmt` if they don't want to be mentioned for daily review task. In addition, we can remove some administrators in `pygmt-contributors` or `pygmt-maintainers` and only keep them in `pygmt-admin/core`  when they don't want to be mentioned for day-to-day review or maintenance tasks.

One question: when should we mention `pygmt-admin/core`? Usually, when contributor/maintainers need admin permission, they can just directly mention one or two active administrators (e.g., @seisman and @weiji14). The situation for discussion among the whole `pygmt-admin/core` team seems rare, right?

**Reply by weiji14** [reply]
> One question: when should we mention `pygmt-admin/core`? Usually, when contributor/maintainers need admin permission, they can just directly mention one or two active administrators (e.g., @seisman and @weiji14). The situation for discussion among the whole `pygmt-admin/core` team seems rare, right?

Well, you could just not mention them :slightly_smiling_face: Having teams is not just about pinging a lot of people easily, it's about giving them the permissions to do the work they need to do to make the project run smoothly. I think this blog post on "The Role of a Maintainer" at https://matthewrocklin.com/blog/2019/05/18/maintainer is worth a read (What we call pygmt-maintainers are akin to 'reviewers' and pygmt-admin/core are akin to actual 'maintainers').

**Reply by core-man** [reply]
Is there any situation that any maintainer only wants to help maintain the project and does not want to be pinged for review/comment requests? In this case, putting `pygmt-maintainers` under `pygmt-contributors` may not be friendly to him/her. This will give Option 1 some consideration.

**Reply by weiji14** [reply]
> pygmt-admin can make any changes and maybe can delete the whole repository smiley

Oh and btw @seisman, I can't delete the GenericMappingTools/pygmt repository as an admin since I'm not a GMT codeowner so stop giving people the wrong ideas :stuck_out_tongue_closed_eyes:

![image](https://user-images.githubusercontent.com/23487320/111291495-2056fa00-86ac-11eb-8f6b-28f6af2cacc6.png)

**Reply by seisman** [reply]
😃 Good to know that not every admin can delete the repo!

**Reply by weiji14** [reply]
Can we arrive at a consensus and finalize the teams restructure?

How about we vote using emojis. Option 1 is :+1:, Option 2 is :eyes: and Option 3 is :rocket: You can vote more than once if either is preferable.

**Reply by maxrjones** [reply]
Sounds good. Let's give people until Sunday night (UTC time) to vote using emojis on @weiji14's post above. Here are the proposed options again so that people don't need to scroll up (but remember, votes go on weiji14's post)
*Option 1*:

```
└── pygmt -  The PyGMT umbrella team
      ├── pygmt-contributors - (write)
      ├── pygmt-maintainers - (maintain)
      └── pygmt-admin/core - (admin)
```

*Option 2*:

```
└── pygmt -  The PyGMT umbrella team
      ├── pygmt-contributors - (write)
      |  └── pygmt-maintainers - (maintain)
      └── pygmt-admin/core - (admin)
```
*Option 3*:
```
└── pygmt -  The PyGMT umbrella team
      └── pygmt-contributors - (write)
        └── pygmt-maintainers - (maintain)
           └── pygmt-admin/core - (admin)
```

**Reply by seisman** [reply]
It seems option 2 is the winner.

**Reply by maxrjones** [reply]
> It seems option 2 is the winner.

Yes, I can format the teams for PyGMT accordingly over the weekend unless you want to do that.

**Reply by maxrjones** [reply]
Changing the names/structure of the existing teams will break past mentions. I think that option is still better than creating a redundant set of teams with the new structure. Please let me know if there are any objections due to that.

**Reply by maxrjones** [reply]
One other comment here - the new structure has `pygmt-contributors` with write permissions. Of the people in `pygmt-contributors` (formerly `python-contributors` - sorry about any broken mentions) who are *not* also in `python-maintainers` or `python`, my opinion is that only @core-man has met the involvement mark for write access. What do people think about renaming python-contributors to pygmt-triage and create a new pygmt-contributors that will include @core-man and pygmt-maintainers as a child team? All those would be under 'pygmt' umbrella team.

**Reply by weiji14** [reply]
I thought the idea was that there could be people in the 'pygmt' umbrella team with 'triage' permissions only (but they won't belong in the other contributors/maintainers/core teams)? I.e. the base set of permissions is 'triage' in team `pygmt`, and `@core-man` would be in `pygmt-contributors` with 'write' permissions, but most others previously in `pygmt-contributors` would just stay in the `pygmt` team?

**Reply by maxrjones** [reply]
OK, mostly wanted to check that I understand who should be in `pygmt` versus `pygmt-contributors`. Functionally, it seems that `python-contributors` will become `pygmt` since that is the triage team. I will update accordingly.

**Reply by weiji14** [reply]
Just to be clear, this is probably the structure we want (correct me if I'm wrong):

```
└── pygmt-team - The PyGMT umbrella team (triage)
      ├── pygmt-contributors - (write)
      |  └── pygmt-maintainers - (maintain)
      └── pygmt-admin - (admin)
```

**Reply by maxrjones** [reply]
> Just to be clear, this is probably the structure we want (correct me if I'm wrong):
> 
> ```
> └── pygmt-team - The PyGMT umbrella team (triage)
>       ├── pygmt-contributors - (write)
>       |  └── pygmt-maintainers - (maintain)
>       └── pygmt-admin - (admin)
> ```

OK.

**Reply by maxrjones** [reply]
Ping @GenericMappingTools/pygmt-contributors

**Comment by maxrjones** [comment]
The current team structure for PyGMT has worked well in my opinion. I will include a discussion about applying the same structure for GMT teams on the agenda for the next community meeting. For reference, here is the PyGMT team structure:
```
└── pygmt-team - The PyGMT umbrella team (triage)
      ├── pygmt-contributors - (write)
      |  └── pygmt-maintainers - (maintain)
      └── pygmt-admin - (admin)
```

**Comment by seisman** [comment]
I'm re-reading the discussions again. The current GMT team structure (https://github.com/orgs/GenericMappingTools/teams) is:

```
|-- gmt-team
    |--- gmt-docs
    |--- gmt-contributors
       |--- gmt-maintainers   
```
I think we should get used to pinging `@GenericMappingTools/gmt-maintainers` for reviewing code changes, ping `@GenericMappingTools/gmt-docs` for reviewing doc-only changes, and ping `@GenericMappingTools/gmt-contributors` when an issue needs more broad discussions.
